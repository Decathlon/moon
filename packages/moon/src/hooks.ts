import * as React from "react";
import {
  QueryObserver,
  QueriesObserver,
  hashQueryKey,
  InfiniteQueryObserver,
  notifyManager,
  QueryKey,
  InfiniteQueryObserverOptions,
  QueryObserverOptions,
  UseQueryResult,
  UseInfiniteQueryResult,
  QueryObserverResult,
  InfiniteData
} from "react-query";
import { Query, QueryState } from "react-query/types/core/query";

import { MoonContext, RquiredMoonContextValue } from "./moonProvider";
import { equal, isServer } from "./utils";

export interface QueriesStates {
  [queryId: string]: QueryState<unknown, unknown> | undefined;
}

export interface ResultProps {
  [propName: string]: any;
}

export interface QueriesResults<Data = any> {
  [queryId: string]: Data | undefined;
}

export type GetState<Data = any, State = Data> = (
  query: Query<Data> | undefined,
  result?: QueryState<Data> | UseQueryResult<Data> | UseInfiniteQueryResult<Data> | undefined,
  isInfinite?: boolean
) => State | undefined;

export function getAdaptedQueryState<Data>(
  query: Query<Data> | undefined,
  result: QueryState<Data> | UseQueryResult<Data> | UseInfiniteQueryResult<Data> | undefined = query?.state,
  isInfinite = false
): QueryState<Data | InfiniteData<Data> | undefined> | undefined {
  if (!query || !result) {
    return undefined;
  }
  const { data: newData, dataUpdatedAt, isFetching, error, errorUpdatedAt, status } = result;
  const networkOnly = query.cacheTime === 0;
  const cachedData = query.state.data;
  const data = newData || cachedData;
  const queryData = !isInfinite && isFetching && networkOnly ? undefined : data;
  return { ...query?.state, dataUpdatedAt, isFetching, error, errorUpdatedAt, status, data: queryData };
}

export function useQueryObserver<State = any, Data = any>(
  queryId: string,
  getState: GetState<Data, State>,
  isInfinite = false
): State | undefined {
  const { store } = useMoon();
  const isMounted = useIsMounted();
  const query = store.getQueryCache().get<Data>(hashQueryKey(queryId));
  const initialSate = getState(query);
  const [state, setState] = React.useState<State | undefined>(initialSate);

  const observerRef = React.useRef<InfiniteQueryObserver<Data> | QueryObserver<Data>>();
  const defaultOptions: InfiniteQueryObserverOptions<Data> | QueryObserverOptions<Data> = React.useMemo(
    () => store.defaultQueryObserverOptions<Data, unknown, Data, Data, QueryKey>({ queryKey: queryId, enabled: false }),
    [queryId]
  );
  const createObserver = React.useCallback(() => {
    return isInfinite
      ? new InfiniteQueryObserver<Data>(store, defaultOptions as InfiniteQueryObserverOptions<Data>)
      : new QueryObserver<Data>(store, defaultOptions);
  }, [isInfinite, store, defaultOptions]);
  const observer = observerRef.current || createObserver();
  observerRef.current = observer;

  if (observer.hasListeners()) {
    // @ts-ignore InfiniteQueryObserver | QueryObserver
    observer.setOptions(defaultOptions);
  }

  const listener = React.useCallback(
    (result: UseQueryResult<Data> | UseInfiniteQueryResult<Data>) => {
      if (isMounted()) {
        const queryState = getState(query, result);
        if (!equal(state || null, queryState || null)) {
          setState(queryState);
        }
      }
    },
    [queryId]
  );

  React.useEffect(() => {
    return observer.subscribe(notifyManager.batchCalls(listener));
  }, [queryId]);

  return state;
}

export function useQueriesObserver<State = any>(queriesIds: string[], getState: GetState<unknown, State>): QueriesResults {
  const { value: currentQueriesIds } = usePrevValue(queriesIds);
  const { store } = useMoon();
  const isMounted = useIsMounted();

  const queriesResults = React.useMemo(() => {
    return currentQueriesIds.reduce<QueriesResults<State>>((result, queryId) => {
      const query = store.getQueryCache().get(hashQueryKey(queryId));
      result[queryId] = getState(query);
      return result;
    }, {});
  }, [currentQueriesIds, store]);

  const [states, setStates] = React.useState<QueriesResults>(queriesResults);

  const queries = React.useMemo(() => {
    return queriesIds.map(queryId =>
      store.defaultQueryObserverOptions<unknown, unknown, unknown, unknown, QueryKey>({ queryKey: queryId, enabled: false })
    );
  }, [currentQueriesIds]);

  const observerRef = React.useRef<QueriesObserver>();
  const observer = observerRef.current || new QueriesObserver(store, queries);
  observerRef.current = observer;

  if (observer.hasListeners()) {
    observer.setQueries(queries);
  }
  const listener = React.useCallback(
    (results: QueryObserverResult[]) => {
      if (isMounted()) {
        queriesIds.forEach((queryId, index) => {
          const query = store.getQueryCache().get(hashQueryKey(queryId));
          const queryState = getState(query, results[index]);
          if (!equal(states[queryId] || null, queryState || null)) {
            setStates({ [queryId]: queryState });
          }
        });
      }
    },
    [currentQueriesIds]
  );

  React.useEffect(() => {
    return observer.subscribe(notifyManager.batchCalls(listener));
  }, [currentQueriesIds]);

  return states;
}

export function useQueryResult<Data = any, Props = ResultProps>(
  queryId: string,
  resultToProps?: (state?: Data | InfiniteData<Data>) => Props,
  isInfinite = false
): Data | InfiniteData<Data> | Props | undefined {
  const getData: GetState<Data, Data | InfiniteData<Data>> = (query, result) => {
    const queryState = getAdaptedQueryState<Data>(query, result);
    return queryState?.data;
  };

  const queryResult = useQueryObserver<Data | InfiniteData<Data>, Data>(queryId, getData, isInfinite);
  return resultToProps ? resultToProps(queryResult) : queryResult;
}

export function useQueriesResults<Props = ResultProps>(
  queriesIds: string[],
  resultsToProps?: (results: QueriesResults) => Props
): QueriesResults | Props {
  const getState: GetState = (query, result) => {
    const queryState = getAdaptedQueryState(query, result);
    return queryState?.data;
  };
  const queriesResult = useQueriesObserver(queriesIds, getState);
  return resultsToProps ? resultsToProps(queriesResult) : queriesResult;
}

export function useQueryState<Data = any, Props = ResultProps>(
  queryId: string,
  stateToProps?: (state?: QueryState<Data | InfiniteData<Data> | undefined, unknown>) => Props,
  isInfinite = false
): QueryState<Data | InfiniteData<Data> | undefined, unknown> | Props | undefined {
  const queryResult = useQueryObserver<QueryState<Data | InfiniteData<Data> | undefined, unknown>, Data>(
    queryId,
    getAdaptedQueryState,
    isInfinite
  );
  return stateToProps ? stateToProps(queryResult) : queryResult;
}

export function useQueriesStates<Props = ResultProps>(
  queriesIds: string[],
  statesToProps?: (states: QueriesStates) => Props
): QueriesStates | Props {
  const queriesResult = useQueriesObserver(queriesIds, getAdaptedQueryState);
  return statesToProps ? statesToProps(queriesResult) : queriesResult;
}

export function usePrevValue<Value = any>(value: Value): { value: Value; prevValue: Value } {
  const valueRef = React.useRef<Value>(value);
  const prevValue = valueRef.current;
  //@ts-ignore prevValue is an object
  if (typeof prevValue === "object" && typeof value === "object" && !equal(prevValue, value, true)) {
    valueRef.current = value;
  }
  return { value: valueRef.current, prevValue };
}

export function useMoon(): RquiredMoonContextValue {
  const moonContext = React.useContext(MoonContext);
  const { client, store } = moonContext;
  if (!client || !store) {
    throw new Error("Invariant Violation: Please wrap the root component in a <MoonProvider>");
  }
  return moonContext as RquiredMoonContextValue;
}

export function useIsMounted(): () => boolean {
  const mountedRef = React.useRef(false);
  const isMounted = React.useCallback(() => mountedRef.current, []);
  const useEffect = isServer ? React.useEffect : React.useLayoutEffect;

  useEffect(() => {
    mountedRef.current = true;
    return () => {
      mountedRef.current = false;
    };
  }, []);

  return isMounted;
}
