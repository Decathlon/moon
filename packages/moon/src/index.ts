export * from "react-query";

export { default as MoonClient } from "./moonClient";
export * from "./moonClient";

export { default as MoonProvider } from "./moonProvider";
export * from "./moonProvider";

export { default as Query } from "./query";
export * from "./query";

export { default as useQuery } from "./useQuery";
export * from "./useQuery";

export { default as usePrefetchQuery } from "./usePrefetchQuery";
export * from "./usePrefetchQuery";

export { default as Mutation } from "./mutation";
export * from "./mutation";

export { default as useMutation } from "./useMutation";
export * from "./useMutation";

export { default as InfiniteQuery } from "./infiniteQuery";
export * from "./infiniteQuery";

export { default as useInfiniteQuery } from "./useInfiniteQuery";
export * from "./useInfiniteQuery";

export * from "./utils";
export * from "./hooks";
