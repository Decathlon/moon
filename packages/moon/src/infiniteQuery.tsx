import { UseInfiniteQueryResult } from "react-query";

import { Nullable } from "./typing";
import useInfiniteQuery, { IInfiniteQueryProps } from "./useInfiniteQuery";

export interface IInfiniteQueryChildrenProps<QueryResponse, QueryError>
  extends Omit<UseInfiniteQueryResult<QueryResponse, QueryError>, "fetchNextPage" | "fetchPreviousPage" | "refetch" | "remove"> {
  actions: Pick<UseInfiniteQueryResult<QueryResponse, QueryError>, "fetchNextPage" | "fetchPreviousPage" | "refetch" | "remove">;
}

export type InfiniteQueryChildren<QueryResponse, QueryError> = (
  props: IInfiniteQueryChildrenProps<QueryResponse, QueryError>
  // eslint-disable-next-line no-undef
) => Nullable<JSX.Element>;

export interface IInfiniteQueryComponentProps<
  QueryVariables,
  QueryPageVariables,
  QueryResponse,
  QueryData,
  QueryError,
  QueryConfig
> extends IInfiniteQueryProps<QueryVariables & QueryPageVariables, QueryResponse, QueryData, QueryError, QueryConfig> {
  children?: InfiniteQueryChildren<QueryData, QueryError>;
}

function InfiniteQuery<
  QueryVariables = any,
  QueryPageVariables = any,
  QueryResponse = any,
  QueryData = QueryResponse,
  QueryError = any,
  QueryConfig = any
>(
  props: IInfiniteQueryComponentProps<QueryVariables, QueryPageVariables, QueryResponse, QueryData, QueryError, QueryConfig>
  // eslint-disable-next-line no-undef
): Nullable<JSX.Element> {
  const { children, ...queryProps } = props;
  const [state, actions] = useInfiniteQuery<
    QueryVariables,
    QueryPageVariables,
    QueryResponse,
    QueryData,
    QueryError,
    QueryConfig
  >(queryProps);
  return children ? children({ ...state, actions }) : null;
}

export default InfiniteQuery;
