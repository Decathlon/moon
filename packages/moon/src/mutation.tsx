import { UseMutationResult } from "react-query";

import { MutateType } from "./moonClient";
import { Nullable } from "./typing";
import useMutation, { IMutationProps } from "./useMutation";

export interface IMutationChildrenProps<MutationVariables = any, MutationResponse = any, MutationError = any>
  extends Omit<
    UseMutationResult<MutationResponse | undefined, MutationError, MutationVariables>,
    "mutate" | "mutateAsync" | "reset"
  > {
  actions: Pick<UseMutationResult<MutationResponse | undefined, MutationError, MutationVariables>, "reset"> & {
    mutate: () => void;
    mutateAsync: () => Promise<MutationResponse>;
  };
}

export type MutationChildren<MutationVariables, MutationResponse, MutationError> = (
  props: IMutationChildrenProps<MutationVariables, MutationResponse, MutationError>
  // eslint-disable-next-line no-undef
) => Nullable<JSX.Element>;

export interface IMutationComponentProps<MutationVariables, MutationResponse, MutationError, MutationConfig>
  extends IMutationProps<MutationVariables, MutationResponse, MutationError, MutationConfig> {
  children?: MutationChildren<MutationVariables, MutationResponse, MutationError>;
}

function Mutation<MutationVariables = any, MutationResponse = any, MutationError = any, MutationConfig = any>(
  props: IMutationComponentProps<MutationVariables, MutationResponse | undefined, MutationError, MutationConfig>
  // eslint-disable-next-line no-undef
): Nullable<JSX.Element> {
  const { children, ...mutationProps } = props;
  const [state, actions] = useMutation<MutationVariables, MutationResponse, MutationError, MutationConfig>(mutationProps);
  return children ? children({ ...state, actions }) : null;
}

Mutation.defaultProps = {
  type: MutateType.Post
};

export default Mutation;
