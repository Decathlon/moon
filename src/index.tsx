export { default as MoonClient } from "./moon-client";
export * from "./moon-client";

export { default as MoonProvider } from "./moon-provider";
export * from "./moon-provider";

export { default as Query } from "./query";
export * from "./query";

export { default as useQuery } from "./query-hook";
export * from "./query-hook";

export { default as Mutation } from "./mutation";
export * from "./mutation";

export { default as useMutation } from "./mutation-hook";
export * from "./mutation-hook";

export * from "./utils";
export * from "./hooks";
export { default as getMoonStore } from "./store";
export * from "./store";
