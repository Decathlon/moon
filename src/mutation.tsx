import * as React from "react";
import { AxiosResponse } from "axios";

import { IMoonContextValue, MoonContext } from "./moonProvider";
import { MutateType } from "./moonClient";

interface IChildren<MutationResponse = any> {
  mutate?: () => void;
  response?: MutationResponse;
  loading?: boolean;
  error?: any;
}

export interface IMutationProps<MutationResponse = any, MutationVariables = any> {
  source: string;
  endPoint: string;
  variables?: MutationVariables;
  type?: MutateType;
  children?: (props: IChildren<MutationResponse>) => React.ReactNode;
  onResponse?: (response: MutationResponse) => void;
  onError?: (error: any) => void;
}

interface IProps<MutationResponse = any, MutationVariables = any>
  extends IMutationProps<MutationResponse, MutationVariables>,
    IMoonContextValue {}

interface IState<MutationResponse = any> {
  loading: boolean;
  error: any;
  response?: MutationResponse;
}

export class DumbMutation<MutationResponse = AxiosResponse<any>, MutationVariables = any> extends React.PureComponent<
  IProps<MutationResponse, MutationVariables>,
  IState<MutationResponse>
> {
  static defaultProps = {
    type: MutateType.Post
  };

  constructor(props: IProps) {
    super(props);

    this.state = {
      loading: false,
      error: null
    };
  }

  private mutate = () => {
    const { client, source, type, endPoint, variables, onResponse, onError } = this.props;

    this.setState({ response: undefined, loading: true, error: null }, async () => {
      try {
        // @ts-ignore API context initialized to null
        const response: MutationResponse | undefined = ((await client.mutate(
          source,
          endPoint,
          type,
          variables
        )) as unknown) as MutationResponse;
        this.setState({ loading: false, response }, () => {
          if (onResponse) {
            onResponse(response);
          }
        });
      } catch (err) {
        this.setState({ error: err, loading: false }, () => {
          if (onError) {
            onError(err);
          }
        });
      }
    });
  };

  public render() {
    const { children } = this.props;
    const { loading, error, response } = this.state;
    return children ? children({ mutate: this.mutate, response, loading, error }) : null;
  }
}

export default class Mutation<MutationResponse = any, MutationVariables = any> extends React.PureComponent<
  IMutationProps<MutationResponse, MutationVariables>
> {
  render() {
    return (
      <MoonContext.Consumer>
        {({ client }) => <DumbMutation<MutationResponse, MutationVariables> client={client} {...this.props} />}
      </MoonContext.Consumer>
    );
  }
}
