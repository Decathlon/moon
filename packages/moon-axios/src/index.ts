import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse, AxiosInterceptorManager } from "axios";
import * as qs from "qs";
import { IInterceptors, ClientInstance } from "@decathlon/moon";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const paramsSerializer = (params: any) => qs.stringify(params, { arrayFormat: "repeat" });

function updateCancelToken(config?: AxiosRequestConfig) {
  if (!config?.cancelToken) {
    const cancelTokenSource = axios.CancelToken.source();
    const cancelToken = cancelTokenSource.token;
    const newConfig = config || {};
    newConfig.cancelToken = cancelToken;
    return () => cancelTokenSource.cancel("Query was cancelled by Moon");
  }
  return undefined;
}

export class MoonAxiosInstance implements ClientInstance {
  instance: AxiosInstance;

  interceptors: {
    request: AxiosInterceptorManager<AxiosRequestConfig>;
    response: AxiosInterceptorManager<AxiosResponse>;
  };

  constructor(config?: AxiosRequestConfig, interceptors?: IInterceptors<AxiosRequestConfig, AxiosResponse>) {
    this.instance = axios.create(config);
    this.interceptors = this.instance.interceptors;
    // use request interceptors
    if (interceptors && interceptors.request && interceptors.request.length > 0 && this.interceptors.request) {
      interceptors.request.forEach(interceptor => this.interceptors.request.use(interceptor.onFulfilled, interceptor.onRejected));
    }

    // use response interceptors
    if (interceptors && interceptors.response && interceptors.response.length > 0 && this.interceptors.response) {
      interceptors.response.forEach(interceptor =>
        this.interceptors.response.use(interceptor.onFulfilled, interceptor.onRejected)
      );
    }
  }

  get<T = any, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig): Promise<R> {
    const cancel = updateCancelToken(config);
    const getPromise = this.instance.get<T, R>(url, config);
    //@ts-ignore
    getPromise.cancel = cancel;
    return getPromise;
  }

  post<T = any, R = AxiosResponse<T>>(url: string, data?: any, config?: AxiosRequestConfig): Promise<R> {
    const cancel = updateCancelToken(config);
    const postPromise = this.instance.post<T, R>(url, data, config);
    //@ts-ignore
    postPromise.cancel = cancel;
    return postPromise;
  }

  delete<T = any, R = AxiosResponse<T>>(url: string, config?: AxiosRequestConfig): Promise<R> {
    const cancel = updateCancelToken(config);
    const deletePromise = this.instance.delete<T, R>(url, config);
    //@ts-ignore
    deletePromise.cancel = cancel;
    return deletePromise;
  }

  put<T = any, R = AxiosResponse<T>>(url: string, data?: any, config?: AxiosRequestConfig): Promise<R> {
    const cancel = updateCancelToken(config);
    const putPromise = this.instance.put<T, R>(url, data, config);
    //@ts-ignore
    putPromise.cancel = cancel;
    return putPromise;
  }

  patch<T = any, R = AxiosResponse<T>>(url: string, data?: any, config?: AxiosRequestConfig): Promise<R> {
    const cancel = updateCancelToken(config);
    const patchPromise = this.instance.patch<T, R>(url, data, config);
    //@ts-ignore
    patchPromise.cancel = cancel;
    return patchPromise;
  }
}

export default function createAxiosClient(
  config?: AxiosRequestConfig,
  interceptors?: IInterceptors<AxiosRequestConfig, AxiosResponse>
): MoonAxiosInstance {
  // create Axios client with custom params serializer
  return new MoonAxiosInstance(
    {
      ...config,
      paramsSerializer: config?.paramsSerializer || paramsSerializer
    },
    interceptors
  );
}
