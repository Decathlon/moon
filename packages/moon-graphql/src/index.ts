import { GraphQLClient } from "graphql-request";
import { RequestInit } from "graphql-request/dist/types.dom";
import { IInterceptors, ClientInstance } from "@decathlon/moon";

export interface GraphQLRequestConfig {
  params?: Record<string, any>;
  options?: RequestInit;
}

export interface GraphQLConfig {
  baseURL: string;
  options?: RequestInit;
}

export class MoonGraphQLInstance implements ClientInstance {
  graphQLinterceptors: IInterceptors<RequestInit, any> | undefined;

  config: GraphQLConfig;

  constructor(config: GraphQLConfig, interceptors?: IInterceptors<RequestInit, any>) {
    this.config = config;
    this.graphQLinterceptors = interceptors;
  }

  useRequestInterceptors(config?: GraphQLRequestConfig): RequestInit {
    let options: RequestInit = { ...this.config.options, ...config?.options };
    this.graphQLinterceptors?.request?.forEach(interceptor => {
      if (interceptor.onFulfilled) {
        options = interceptor.onFulfilled(options) as RequestInit;
      }
    });
    return options;
  }

  useResponseInterceptors(response: unknown, isRejected = false): void {
    this.graphQLinterceptors?.response?.forEach(interceptor => {
      if (!isRejected && interceptor.onFulfilled) {
        interceptor.onFulfilled(response);
      }
      if (isRejected && interceptor.onRejected) {
        interceptor.onRejected(response);
      }
    });
  }

  query<Response = any>(query: string, config?: GraphQLRequestConfig): Promise<Response> {
    const options = this.useRequestInterceptors(config);
    const controller = new AbortController();
    if (!options.signal) {
      options.signal = controller.signal;
    }
    const instance = new GraphQLClient(this.config.baseURL, options);
    const getPromise = instance.request<Response, any>(query, config?.params, options.headers).then(
      result => {
        this.useResponseInterceptors(result);
        return result;
      },
      error => {
        this.useResponseInterceptors(error);
        throw error;
      }
    );
    //@ts-ignore
    getPromise.cancel = () => controller.abort();
    return getPromise;
  }

  get<Response = any>(query: string, config?: GraphQLRequestConfig): Promise<Response> {
    return this.query<Response>(query, config);
  }

  post<Response = any>(mutation: string, config?: GraphQLRequestConfig): Promise<Response> {
    return this.query<Response>(mutation, config);
  }

  delete<Response = any>(mutation: string, config?: GraphQLRequestConfig): Promise<Response> {
    return this.query<Response>(mutation, config);
  }

  put<Response = any>(mutation: string, config?: GraphQLRequestConfig): Promise<Response> {
    return this.query<Response>(mutation, config);
  }

  patch<Response = any>(mutation: string, config?: GraphQLRequestConfig): Promise<Response> {
    return this.query<Response>(mutation, config);
  }
}

export default function createAxiosClient(
  config: GraphQLConfig,
  interceptors?: IInterceptors<RequestInit, any>
): MoonGraphQLInstance {
  // create Axios client with custom params serializer
  return new MoonGraphQLInstance(config, interceptors);
}
