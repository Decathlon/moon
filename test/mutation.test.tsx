/* eslint-disable  import/no-extraneous-dependencies */
/// <reference path="./typings/tests-entry.d.ts" />

import * as React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { DumbMutation, IMutationChildrenProps } from "../src/mutation";

describe("Mutation component", () => {
  it("should call children with the mutation state", () => {
    const props = {
      actions: { mutate: jest.fn(), cancel: jest.fn() },
      loading: false,
      error: null,
      response: { foo: "bar" },
      children: (props: IMutationChildrenProps<{ foo: string }>) => <div>{JSON.stringify(props)}</div>
    };
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<DumbMutation {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});
