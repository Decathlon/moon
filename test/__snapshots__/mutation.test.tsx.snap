// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation component should call children with the mutation state 1`] = `
<div>
  {"actions":{},"loading":false,"error":null,"response":{"foo":"bar"}}
</div>
`;
